core/deployment/src/main/java/io/quarkus/deployment/ide/IDEDetectionProcessor.java

```diff
@@ -0,0 +1,116 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.pkg.builditem.BuildSystemTargetBuildItem;
+import io.quarkus.runtime.util.JavaVersionUtil;
+
+public class IDEDetectionProcessor {
+
+    private static Map<String, List<IDE>> IDE_MARKER_FILES = new HashMap<>();
+    private static Map<Predicate<ProcessInfo>, IDE> IDE_PROCESSES = new HashMap<>();
+
+    static {
+        IDE_MARKER_FILES.put(".idea", Collections.singletonList(IDE.IDEA));
+        IDE_MARKER_FILES.put(".project", Arrays.asList(IDE.VSCODE, IDE.ECLIPSE));
+        IDE_MARKER_FILES = Collections.unmodifiableMap(IDE_MARKER_FILES);
+
+        IDE_PROCESSES.put((processInfo -> processInfo.getCommand().contains("idea.")), IDE.IDEA);
+        IDE_PROCESSES.put((processInfo -> processInfo.getCommand().contains("vscode")), IDE.VSCODE);
+        IDE_PROCESSES.put((processInfo -> processInfo.getCommand().contains("eclipse")), IDE.ECLIPSE); // TODO: I have no idea if this is correct
+        IDE_PROCESSES = Collections.unmodifiableMap(IDE_PROCESSES);
+    }
+
+    @BuildStep
+    public IDEFileBuildItem detectIDEFiles(BuildSystemTargetBuildItem buildSystemTarget) {
+        Set<IDE> result = new HashSet<>(2);
+        Path projectRoot = buildSystemTarget.getOutputDirectory().getParent();
+        IDE_MARKER_FILES.forEach((file, ides) -> {
+            if (Files.exists(projectRoot.resolve(file))) {
+                result.addAll(ides);
+            }
+        });
+        return new IDEFileBuildItem(result);
+    }
+
+    @BuildStep
+    public IDERunningProcessBuildItem detectRunningIDEProcesses() {
+        Set<IDE> result = new HashSet<>(2);
+        List<ProcessInfo> processInfos = ProcessUtil.runningProcesses();
+        for (ProcessInfo processInfo : processInfos) {
+            for (Map.Entry<Predicate<ProcessInfo>, IDE> entry : IDE_PROCESSES.entrySet()) {
+                if (entry.getKey().test(processInfo)) {
+                    result.add(entry.getValue());
+                    break;
+                }
+            }
+        }
+        return new IDERunningProcessBuildItem(result);
+    }
+
+    // TODO: remove when we move to Java 11 and just call the methods of 'java.lang.ProcessHandle'
+    private static class ProcessUtil {
+
+        /**
+         * Returns a list of running processes
+         * Only works for Java 11+
+         */
+        public static List<ProcessInfo> runningProcesses() {
+            if (!JavaVersionUtil.isJava11OrHigher()) {
+                return Collections.emptyList();
+            }
+            // we can't use ProcessHandle directly as it is Java 9+, so we need to do reflection to the get the info we need
```

Comment:
Use `io.smallrye.common:smallrye-common-os`, which already does that in MR JARs: 

https://github.com/smallrye/smallrye-common/blob/master/os/src/main/java/io/smallrye/common/os/Process.java

Reply:
Oh, cool!

Reply:
But it doen't look like that piece of code gives you information about **all** the running processes, so I don't think I can use it here

Reply:
Maybe we can easily add that to the library if @dmlloyd agrees :)

Reply:
I created https://github.com/smallrye/smallrye-common/issues/79 just in case